File: src/main/python/AoC2020_05.py
#! /usr/bin/env python3
#
# Advent of Code 2020 Day 5
#
from aoc import my_aocd
import aocd


def _parse(inputs: tuple[str]) -> list[str]:
    table = str.maketrans("FBLR", "0101")
    return sorted(line.translate(table) for line in inputs)


def _as_int(binary: str) -> int:
    return int(binary, 2)


def part_1(inputs: tuple[str]) -> int:
    translated = _parse(inputs)
    return _as_int(translated[-1])


def part_2(inputs: tuple[str]) -> int:
    translated = _parse(inputs)
    for i in range(len(translated) - 1):
        if translated[i][-1] == translated[i+1][-1]:
            return _as_int(translated[i]) + 1
    raise ValueError("Unsolvable")


TEST1 = """\
FBFBBFFRLR
BFFFBBFRRR
FFFBBBFRRR
BBFFBBFRLL
""".splitlines()


TEST2 = """\
FFFFFFFLLL
FFFFFFFLLR
FFFFFFFLRL
FFFFFFFRLL
FFFFFFFRLR
""".splitlines()


def main() -> None:
    puzzle = aocd.models.Puzzle(2020, 5)
    my_aocd.print_header(puzzle.year, puzzle.day)

    assert part_1(TEST1) == 820
    assert part_2(TEST2) == 3

    inputs = my_aocd.get_input(2020, 5, 839)
    result1 = part_1(inputs)
    print(f"Part 1: {result1}")
    result2 = part_2(inputs)
    print(f"Part 2: {result2}")
    my_aocd.check_results(puzzle, result1, result2)


if __name__ == '__main__':
    main()

File: src/main/julia/AoC2020_05.jl
#! /usr/bin/env julia
#

include("aoc_main.jl")
if abspath(PROGRAM_FILE) == @__FILE__
    include("aocd.jl")
    using .Aocd
end

module AoC2020_05

include("aoc.jl")

function _parse(input)
    table = Dict('F' => '0', 'B' => '1', 'L' => '0', 'R' => '1')
    return sort([join([table[c] for c ∈ line]) for line ∈ input])
end

function asInt(string)
    return parse(Int, string, base=2)
end

function part1(input)
    translated = _parse(input)
    return asInt(last(translated))
end

function part2(input)
    translated = _parse(input)
    for i ∈ 2:length(translated)
        if last(translated[i]) == last(translated[i-1])
            return asInt(translated[i]) - 1
        end
    end
    throw(error("Unsolvable"))
end

TEST1 = @aoc_splitlines("""
FBFBBFFRLR
BFFFBBFRRR
FFFBBBFRRR
BBFFBBFRLL
""")
TEST2 = @aoc_splitlines("""
FFFFFFFLLL
FFFFFFFLLR
FFFFFFFLRL
FFFFFFFRLL
FFFFFFFRLR
""")

function samples()
    @assert part1(TEST1) == 820
    @assert part2(TEST2) == 3
end

end # module AoC2020_05

aoc_main(@__FILE__, ARGS, 2020, 5)
